File: apps/ios/BananaTape/BananaTape/EditorModels.swift
import Foundation

struct EditorPoint: Equatable {
    let x: Double
    let y: Double
}

struct EditorSize: Equatable {
    let width: Double
    let height: Double
}

enum EditorProvider: String, Equatable {
    case mock
    case openAI = "openai"
    case codex = "god-tibo"
}

enum ImageGenerationStatus: String, Equatable {
    case pending
    case ready
    case error
}

enum EditorMode: String, Equatable {
    case generate
    case edit
}

enum OutputSize: String, Equatable {
    case square = "1024x1024"
    case portrait = "1024x1536"
    case landscape = "1536x1024"
}

enum AnnotationStatus: String, Equatable {
    case pending
    case review
    case accepted
}

enum DrawingTool: String, Equatable {
    case pen
    case arrow
}

struct DrawingPath: Equatable {
    let id: String
    let tool: DrawingTool
    let points: [EditorPoint]
    let color: String
    let strokeWidth: Double
}

struct BoundingBox: Equatable {
    let id: String
    let x: Double
    let y: Double
    let width: Double
    let height: Double
    let color: String
    let status: AnnotationStatus
}

struct TextMemo: Equatable {
    let id: String
    let x: Double
    let y: Double
    let text: String
    let color: String
}

struct CanvasAnnotations: Equatable {
    let paths: [DrawingPath]
    let boxes: [BoundingBox]
    let memos: [TextMemo]

    static let empty = CanvasAnnotations(paths: [], boxes: [], memos: [])
}

struct CanvasImage: Equatable {
    let id: String
    let url: String
    let assetId: String?
    let size: EditorSize
    let position: EditorPoint
    let parentId: String?
    let generationIndex: Int
    let generationBatchId: String?
    let batchIndex: Int?
    let prompt: String
    let provider: EditorProvider
    let mode: EditorMode
    let createdAt: Double
    let annotations: CanvasAnnotations
    let hasMagicLayerFields: Bool
    let status: ImageGenerationStatus
    let userErrorMessage: String?

    init(id: String, url: String, assetId: String?, size: EditorSize, position: EditorPoint, parentId: String?, generationIndex: Int, generationBatchId: String? = nil, batchIndex: Int? = nil, prompt: String, provider: EditorProvider, mode: EditorMode, createdAt: Double, annotations: CanvasAnnotations, hasMagicLayerFields: Bool, status: ImageGenerationStatus, userErrorMessage: String?) {
        self.id = id
        self.url = url
        self.assetId = assetId
        self.size = size
        self.position = position
        self.parentId = parentId
        self.generationIndex = generationIndex
        self.generationBatchId = generationBatchId
        self.batchIndex = batchIndex
        self.prompt = prompt
        self.provider = provider
        self.mode = mode
        self.createdAt = createdAt
        self.annotations = annotations
        self.hasMagicLayerFields = hasMagicLayerFields
        self.status = status
        self.userErrorMessage = userErrorMessage
    }

    var canEditMagicLayers: Bool { false }
    static let magicLayerEditingMessage = "Magic Layer editing is desktop-only"
}

struct EditorState: Equatable {
    let provider: EditorProvider
    let mode: EditorMode
    let outputSize: OutputSize
    let focusedImageIds: [String]
}

File: apps/android/app/src/main/java/app/bananatape/mobile/editor/EditorModels.kt
package app.bananatape.mobile.editor

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

data class EditorPoint(val x: Double, val y: Double)

data class EditorSize(val width: Double, val height: Double)

enum class EditorProvider(val value: String) {
    MOCK("mock"),
    OPENAI("openai"),
    CODEX("god-tibo"),
}

enum class ImageGenerationStatus(val value: String) {
    PENDING("pending"),
    READY("ready"),
    ERROR("error"),
}

enum class EditorMode(val value: String) {
    GENERATE("generate"),
    EDIT("edit"),
}

enum class LineageDirection {
    LEFT,
    RIGHT,
    UP,
    DOWN,
}

fun lineageSwipeDirection(
    tool: CanvasTool,
    deltaX: Float,
    deltaY: Float,
    minimumDistance: Float = 48f,
    axisDominanceRatio: Float = 1.25f,
): LineageDirection? {
    if (tool != CanvasTool.SELECT) return null
    val horizontalDistance = abs(deltaX)
    val verticalDistance = abs(deltaY)
    val dominantDistance = max(horizontalDistance, verticalDistance)
    val secondaryDistance = min(horizontalDistance, verticalDistance)
    if (dominantDistance < minimumDistance || dominantDistance < secondaryDistance * axisDominanceRatio) return null
    return if (horizontalDistance > verticalDistance) {
        if (deltaX < 0f) LineageDirection.RIGHT else LineageDirection.LEFT
    } else {
        if (deltaY < 0f) LineageDirection.DOWN else LineageDirection.UP
    }
}

fun ComposerState.withFocusedImageSelection(image: CanvasImage?): ComposerState =
    copy(
        hasSelectedImage = image?.status == ImageGenerationStatus.READY,
        mode = if (image?.status == ImageGenerationStatus.READY) EditorMode.EDIT else EditorMode.GENERATE,
    )

fun ComposerState.openingForFocusedImage(image: CanvasImage?): ComposerState =
    withFocusedImageSelection(image)

fun ComposerState.startingNewGeneration(): ComposerState =
    copy(mode = EditorMode.GENERATE)

fun resolvedSubmissionMode(composerMode: EditorMode, focusedImage: CanvasImage?): EditorMode =
    if (focusedImage?.status == ImageGenerationStatus.READY) composerMode else EditorMode.GENERATE

enum class OutputSize(val value: String) {
    SQUARE("1024x1024"),
    PORTRAIT("1024x1536"),
    LANDSCAPE("1536x1024"),
}

enum class AnnotationStatus(val value: String) {
    PENDING("pending"),
    REVIEW("review"),
    ACCEPTED("accepted"),
}

enum class DrawingTool(val value: String) {
    PEN("pen"),
    ARROW("arrow"),
}

data class DrawingPath(
    val id: String,
    val tool: DrawingTool,
    val points: List<EditorPoint>,
    val color: String,
    val strokeWidth: Double,
)

data class BoundingBox(
    val id: String,
    val x: Double,
    val y: Double,
    val width: Double,
    val height: Double,
    val color: String,
    val status: AnnotationStatus,
)

data class TextMemo(
    val id: String,
    val x: Double,
    val y: Double,
    val text: String,
    val color: String,
)

data class CanvasAnnotations(
    val paths: List<DrawingPath>,
    val boxes: List<BoundingBox>,
    val memos: List<TextMemo>,
) {
    companion object {
        val Empty = CanvasAnnotations(paths = emptyList(), boxes = emptyList(), memos = emptyList())
    }
}

data class CanvasImage(
    val id: String,
    val url: String,
    val assetId: String?,
    val size: EditorSize,
    val position: EditorPoint,
    val parentId: String?,
    val generationIndex: Int,
    val prompt: String,
    val provider: EditorProvider,
    val mode: EditorMode,
    val createdAt: Double,
    val annotations: CanvasAnnotations,
    val hasMagicLayerFields: Boolean,
    val status: ImageGenerationStatus = ImageGenerationStatus.READY,
    val userErrorMessage: String? = null,
    val generationBatchId: String? = null,
    val batchIndex: Int? = null,
) {
    val canEditMagicLayers: Boolean = false

    companion object {
        const val MagicLayerEditingMessage = "Magic Layer editing is desktop-only"
    }
}

data class EditorState(
    val provider: EditorProvider,
    val mode: EditorMode,
    val outputSize: OutputSize,
    val focusedImageIds: List<String>,
)
